// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2019 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

syntax = "proto2";

package recidiviz.ingest.models;

// To regenerate, ensure `mypy-protobuf` is installed:
// ```bash
// pipenv install --dev
// ```
// See the README for instructions on installing protoc if you do not yet have
// it installed.
//
// Re-generate the python files including type information:
// ```bash
// protoc recidiviz/ingest/models/ingest_info.proto --python_out . --mypy_out .
// ```

message IngestInfo {
    // TODO(#8905): Delete this whole file once ingest mappings overhaul is complete for all states.

    repeated StatePerson state_people = 9;
    repeated StatePersonRace state_person_races = 10;
    repeated StatePersonEthnicity state_person_ethnicities = 11;
    repeated StateAlias state_aliases = 12;
    repeated StatePersonExternalId state_person_external_ids = 13;
    repeated StateAssessment state_assessments = 14;
    repeated StateSupervisionSentence state_supervision_sentences = 16;
    repeated StateIncarcerationSentence state_incarceration_sentences = 17;
    repeated StateCharge state_charges = 19;
    repeated StateCourtCase state_court_cases = 20;
    repeated StateIncarcerationPeriod state_incarceration_periods = 22;
    repeated StateSupervisionPeriod state_supervision_periods = 23;
    repeated StateSupervisionCaseTypeEntry state_supervision_case_type_entries = 24;
    repeated StateIncarcerationIncident state_incarceration_incidents = 25;
    repeated StateIncarcerationIncidentOutcome state_incarceration_incident_outcomes = 26;
    repeated StateSupervisionViolation state_supervision_violations = 28;
    repeated StateSupervisionViolationTypeEntry state_supervision_violation_type_entries = 29;
    repeated StateSupervisionViolatedConditionEntry state_supervision_violated_condition_entries = 30;
    repeated StateSupervisionViolationResponse state_supervision_violation_responses = 31;
    repeated StateSupervisionViolationResponseDecisionEntry state_supervision_violation_response_decision_entries = 32;
    repeated StateAgent state_agents = 33;
    repeated StateProgramAssignment state_program_assignments = 34;
    repeated StateEarlyDischarge state_early_discharges = 35;
    repeated StateSupervisionContact state_supervision_contacts = 36;
}

message StatePerson {
    optional string state_person_id = 1;
    optional string full_name = 2;
    optional string surname = 3;
    optional string given_names = 4;
    optional string middle_names = 5;
    optional string name_suffix = 6;
    optional string birthdate = 7;
    optional string gender = 8;
    optional string age = 9;
    optional string current_address = 10;
    optional string residency_status = 11;

    repeated string state_person_race_ids = 12;
    repeated string state_person_ethnicity_ids = 13;
    repeated string state_alias_ids = 14;
    repeated string state_person_external_ids_ids = 15;
    repeated string state_assessment_ids = 16;
    repeated string state_supervision_sentence_ids = 26;
    repeated string state_incarceration_sentence_ids = 27;
    repeated string state_incarceration_period_ids = 24;
    repeated string state_supervision_period_ids = 25;
    repeated string state_program_assignment_ids = 18;
    repeated string state_incarceration_incident_ids = 21;
    repeated string state_supervision_violation_ids = 22;
    repeated string state_supervision_contact_ids = 23;
    optional string supervising_officer_id = 19;
    optional string state_code = 20;
}

message StatePersonRace {
    optional string state_person_race_id = 1;
    optional string race = 2;
    optional string state_code = 3;
}

message StatePersonEthnicity {
    optional string state_person_ethnicity_id = 1;
    optional string ethnicity = 2;
    optional string state_code = 3;
}

message StateAlias {
    optional string state_alias_id = 1;
    optional string state_code = 2;
    optional string full_name = 3;
    optional string surname = 4;
    optional string given_names = 5;
    optional string middle_names = 6;
    optional string name_suffix = 7;
    optional string alias_type = 8;
}

message StatePersonExternalId {
    optional string state_person_external_id_id = 1;
    optional string id_type = 2;
    optional string state_code = 3;
}

message StateAssessment {
    optional string state_assessment_id = 1;
    optional string assessment_class = 2;
    optional string assessment_type = 3;
    optional string assessment_date = 4;
    optional string state_code = 5;
    optional string assessment_score = 6;
    optional string assessment_level = 7;
    optional string assessment_metadata = 8;

    optional string conducting_agent_id = 9;
}

message StateSupervisionSentence {
    optional string state_supervision_sentence_id = 1;
    optional string status = 2;
    optional string supervision_type = 3;
    optional string date_imposed = 4;
    optional string start_date = 5;
    optional string projected_completion_date = 6;
    optional string completion_date = 7;
    optional string state_code = 8;
    optional string county_code = 9;
    optional string min_length = 10;
    optional string max_length = 11;

    repeated string state_charge_ids = 12;
    repeated string state_early_discharge_ids = 15;
}

message StateIncarcerationSentence {
    optional string state_incarceration_sentence_id = 1;
    optional string status = 2;
    optional string incarceration_type = 3;
    optional string date_imposed = 4;
    optional string start_date = 5;
    optional string projected_min_release_date = 6;
    optional string projected_max_release_date = 7;
    optional string completion_date = 8;
    optional string parole_eligibility_date = 9;
    optional string state_code = 10;
    optional string county_code = 11;
    optional string min_length = 12;
    optional string max_length = 13;
    optional string is_life = 14;
    optional string is_capital_punishment = 15;
    optional string parole_possible = 16;
    optional string initial_time_served = 17;
    optional string good_time = 18;
    optional string earned_time = 19;

    repeated string state_charge_ids = 20;
    repeated string state_early_discharge_ids = 23;
}

message StateCharge {
    optional string state_charge_id = 1;
    optional string status = 2;
    optional string offense_date = 3;
    optional string date_charged = 4;
    optional string state_code = 5;
    optional string county_code = 6;
    optional string ncic_code = 7;
    optional string statute = 8;
    optional string description = 9;
    optional string attempted = 10;
    optional string classification_type = 11;
    optional string classification_subtype = 12;
    optional string offense_type = 19;
    optional string is_violent = 20;
    optional string is_sex_offense = 21;
    optional string counts = 13;
    optional string charge_notes = 14;
    optional string is_controlling = 15;
    optional string charging_entity = 16;

    optional string state_court_case_id = 17;
}

message StateCourtCase {
    optional string state_court_case_id = 1;
    optional string status = 2;
    optional string court_type = 3;
    optional string date_convicted = 4;
    optional string next_court_date = 5;
    optional string state_code = 6;
    optional string county_code = 7;
    optional string judicial_district_code = 8;

    optional string judge_id = 10;
}

message StateIncarcerationPeriod {
    optional string state_incarceration_period_id = 1;
    optional string incarceration_type = 3;
    optional string admission_date = 4;
    optional string release_date = 5;
    optional string state_code = 6;
    optional string county_code = 7;
    optional string facility = 8;
    optional string housing_unit = 9;
    optional string admission_reason = 11;
    optional string release_reason = 13;
    optional string specialized_purpose_for_incarceration = 14;
    optional string custodial_authority = 15;
}

message StateSupervisionPeriod {
    optional string state_supervision_period_id = 1;
    optional string supervision_type = 4;
    optional string start_date = 5;
    optional string termination_date = 6;
    optional string state_code = 7;
    optional string county_code = 8;
    optional string supervision_site = 9;
    optional string admission_reason = 10;
    optional string termination_reason = 11;
    optional string supervision_level = 12;
    optional string supervising_officer_id = 13;
    optional string custodial_authority = 14;
    optional string conditions = 15;

    repeated string state_supervision_case_type_entry_ids = 19;
}

message StateSupervisionCaseTypeEntry {
    optional string state_supervision_case_type_entry_id = 1;
    optional string case_type = 2;
    optional string state_code = 3;
}

message StateIncarcerationIncident {
    optional string state_incarceration_incident_id = 1;
    optional string incident_type = 2;
    optional string incident_date = 3;
    optional string state_code = 4;
    optional string facility = 5;
    optional string location_within_facility = 6;
    optional string incident_details = 7;

    optional string responding_officer_id = 8;
    repeated string state_incarceration_incident_outcome_ids = 9;
}

message StateIncarcerationIncidentOutcome {
    optional string state_incarceration_incident_outcome_id = 1;
    optional string outcome_type = 2;
    optional string date_effective = 3;
    optional string hearing_date = 4;
    optional string report_date = 5;
    optional string state_code = 6;
    optional string outcome_description = 7;
    optional string punishment_length_days = 8;
}

message StateSupervisionViolatedConditionEntry {
    optional string state_supervision_violated_condition_entry_id = 1;
    optional string condition = 2;
    optional string state_code = 3;
}

message StateSupervisionViolationTypeEntry {
    optional string state_supervision_violation_type_entry_id = 1;
    optional string violation_type = 2;
    optional string state_code = 3;
}

message StateSupervisionViolation {
    optional string state_supervision_violation_id = 1;
    optional string violation_date = 3;
    optional string state_code = 4;
    optional string is_violent = 5;
    optional string is_sex_offense = 6;

    repeated string state_supervision_violation_response_ids = 8;
    repeated string state_supervision_violated_condition_entry_ids = 9;
    repeated string state_supervision_violation_type_entry_ids = 10;
}

message StateSupervisionViolationResponseDecisionEntry {
    optional string state_supervision_violation_response_decision_entry_id = 1;
    optional string decision = 2;
    optional string state_code = 4;
}

message StateSupervisionViolationResponse {
    optional string state_supervision_violation_response_id = 1;
    optional string response_type = 2;
    optional string response_subtype = 3;
    optional string response_date = 4;
    optional string state_code = 5;
    optional string deciding_body_type = 8;
    optional string is_draft = 9;

    repeated string decision_agent_ids = 10;
    repeated string state_supervision_violation_response_decision_entry_ids = 11;
}

message StateAgent {
    optional string state_agent_id = 1;
    optional string agent_type = 2;
    optional string state_code = 3;
    optional string full_name = 4;
    optional string surname = 5;
    optional string given_names = 6;
    optional string middle_names = 7;
    optional string name_suffix = 8;
}

message StateProgramAssignment{
    optional string state_program_assignment_id = 1;
    optional string participation_status = 2;
    optional string referral_date = 3;
    optional string start_date = 4;
    optional string discharge_date = 5;
    optional string state_code = 6;
    optional string program_id = 7;
    optional string program_location_id = 8;
    optional string referral_metadata = 11;

    optional string referring_agent_id = 10;
}

message StateEarlyDischarge{
    optional string state_early_discharge_id = 1;
    optional string request_date = 2;
    optional string decision_date = 3;
    optional string decision = 4;
    optional string deciding_body_type = 5;
    optional string requesting_body_type = 6;
    optional string state_code = 7;
    optional string county_code = 8;
    optional string decision_status = 9;
}

message StateSupervisionContact {
     optional string state_supervision_contact_id = 1;
     optional string contact_date = 2;
     optional string contact_reason = 3;
     optional string contact_type = 4;
     optional string contact_method = 11;
     optional string location = 5;
     optional string resulted_in_arrest = 6;
     optional string status = 7;
     optional string verified_employment = 8;
     optional string state_code = 9;

     optional string contacted_agent_id = 10;
}
